### PR TITLE
Migrate and complete methods block documentation

### PR DESCRIPTION
**Pull request by mdgeorge4153**
This PR rewrites most of the existing documentation on the methods block and summarization and merges it into the reference manual.

### CODE REVIEW

**Comment by mdgeorge4153** [comment]
Generated documentation is here: https://certora-certora-prover-documentation--21.com.readthedocs.build/en/21/docs/ref-manual/cvl/methods.html

File: docs/ref-manual/cvl/methods.md [hunk 1]
```diff
@@ -46,44 +55,280 @@ method_summary   ::= "ALWAYS" "(" value ")"
...
-                   | [ "with" "(" cvl_params ")" ] expression
+                   | [ "with" "(" "env" id ")" ] block
+                   | [ "with" "(" "env" id ")" ] expression
+                   | [ "with" "(" "env" id ")" ] id "(" [ id { "," id } ] ")"
 
-cvl_param ::= cvl_type [ id ]
+```
 
+```{todo}
+The `with(env id) block` and `with(env id) expression` productions come from
```
Comment (thomasbernardi) [inline docs/ref-manual/cvl/methods.md:65]: The only valid expression summaries are applications of ghosts and cvl functions (I do not recall if we allow definitions but I suspect not)

File: docs/ref-manual/cvl/methods.md [hunk 2]
```diff
@@ -46,44 +55,280 @@ method_summary   ::= "ALWAYS" "(" value ")"
...
+in the contract being verified does not depend on any of the environment
+variables.  The results of this check are displayed on the verification report
+as a separate rule called `envfreeFuncsStaticCheck`.
+
 ```{todo}
-This document is incomplete.  See {doc}`/docs/confluence/advanced/methods-overview`,
-{doc}`/docs/confluence/advanced/summaries`, {doc}`/docs/confluence/advanced/internal-summaries`,
-and {doc}`/docs/confluence/advanced/expressive-summaries` for the old documentation
-about the methods block.
+There is a separate check called `envfreeFuncsAreNonpayable`.  Why is this
```
Comment (thomasbernardi) [inline docs/ref-manual/cvl/methods.md:105]: I always assumed the "static check" didn't require a call to z3 while the other envfree rules did?

File: docs/ref-manual/cvl/methods.md [hunk 3]
```diff
@@ -46,44 +55,280 @@ method_summary   ::= "ALWAYS" "(" value ")"
...
-### Application policies (UNRESOLVED or ALL)
+ * If the code for the function is not known at verification time, then the
+   function call must be summarized.  If no summary is given, the default summary
+   type is {ref}`AUTO <auto-summary>`, whose behavior is determined by the type of
+   function call.  In this case, the verification report will contain a contract
+   call resolution warning.
 
-### `ALWAYS`, `CONSTANT`, `PER_CALLEE_CONSTANT`
+```{todo}
+The default application policy is currently undocumented.
```
Comment (thomasbernardi) [inline docs/ref-manual/cvl/methods.md:152]: The default policy is `ALL` with the exception of a dispatcher summary whose default policy is `UNRESOLVED`.

File: docs/ref-manual/cvl/methods.md [hunk 4]
```diff
@@ -46,44 +55,280 @@ method_summary   ::= "ALWAYS" "(" value ")"
...
+```
+
+```{todo}
+The old documentation is ambiguous about the behavior of `UNRESOLVED` summaries
```
Comment (thomasbernardi) [inline docs/ref-manual/cvl/methods.md:156]: It is impossible to have an "unresolved" internal method, so internal methods will only be summarized if the resolution type is `ALL` (since the code for the function is known at verification time).

That being said, obviously if the internal call is part of an external call that was not resolved, we will know nothing about the internal function call (but the external call will already be summarized)

File: docs/ref-manual/cvl/methods.md [hunk 5]
```diff
@@ -46,44 +55,280 @@ method_summary   ::= "ALWAYS" "(" value ")"
...
+
+ * The `NONDET` approximation makes no assumptions about the return values; each
+   call to the summarized method may return a different result.
+
+```{todo}
+The old documentation seems to imply that the ETH balances for calls to
+`ALWAYS`, `CONSTANT`, and `PER_CALLEE_CONSTANT` do not account for the ETH value
+passed in the function call, while `NONDET` does.
+
+This seems like the wrong behavior?
```
Comment (thomasbernardi) [inline docs/ref-manual/cvl/methods.md:193]: I mean, aren't all of the summaries "wrong" behavior?

But more seriously, what would it even mean to "account for the ETH value passed in the function call" even mean when you are replacing a function call with a either a pure function which returns a static value, or a havoc'd value (ALWAYS and CONSTANT) or a havoc'd value per callee (PER_CALLEE_CONSTANT). Like, these are unsound summarizations which just replace the entire function and know _nothing_ about the parameters or state or anything? Also, how does NONDET account for the ETH value??

File: docs/ref-manual/cvl/methods.md [hunk 6]
```diff
@@ -46,44 +55,280 @@ method_summary   ::= "ALWAYS" "(" value ")"
...
+```
+
+```{todo}
+The following note from the old documentation needs clarification:
```
Comment (thomasbernardi) [inline docs/ref-manual/cvl/methods.md:197]: @jtoman ?

File: docs/ref-manual/cvl/methods.md [hunk 7]
```diff
@@ -46,44 +55,280 @@ method_summary   ::= "ALWAYS" "(" value ")"
...
+ * Normal calls and constructors use the `HAVOC_ECF` approximation: they are
+   assumed to change the state of external contracts arbitrarily but to leave
+   the caller's state unchanged.
+
+ * Calls to library methods are assumed to change the caller's state in an
+   arbitrary way, but are assumed to leave the state of other contracts
+   unchanged.
+
+```{todo}
+The effect of library calls on the current contract's ETH balance is unclear.
```
Comment (thomasbernardi) [inline docs/ref-manual/cvl/methods.md:290]: @jtoman

File: docs/ref-manual/cvl/methods.md [hunk 8]
```diff
@@ -46,44 +55,280 @@ method_summary   ::= "ALWAYS" "(" value ")"
...
+variables are not supported.
+
+Functions used as summaries are not allowed to call contract functions.  They
+may only accept parameter types that are expressible in solidity; extended CVL
+types like `method` and `mathint` cannot be passed as arguments.
+
+### Undocumented summaries
+
+```{todo}
+Block summaries are undocumented.  Expression summaries other than function
```
Comment (thomasbernardi) [inline docs/ref-manual/cvl/methods.md:330]: block summaries don't exist
expression summaries other than function summaries don't exist
don't let the parser fool you :P

File: docs/ref-manual/cvl/methods.md (hunk 1 above)
Comment (thomasbernardi) [reply docs/ref-manual/cvl/methods.md:65]: additionally, `with (env e)` does nothing and should not be used in summaries

File: docs/ref-manual/cvl/methods.md (hunk 2 above)
Comment (mdgeorge4153) [reply docs/ref-manual/cvl/methods.md:105]: The question is why do we need a check for nonpayability?

File: docs/ref-manual/cvl/methods.md (hunk 6 above)
Comment (jtoman) [reply docs/ref-manual/cvl/methods.md:197]: Yeah, this is pretty in the weeds, but the amount of data returned by the caller does not necessarily need to honor the request of the caller (in fact, in the case of dynamically sized return payloads, there is no "up front" size that the caller can declare for the expected size of the return buffer).

Instead, there is a sequence of sanity checks in the caller to establish that the callee returned enough data to satisfy the expected return values. So, if contract A calls contract B expecting contract B to return two uint256 values, and contract B only returns 32 bytes, then the call within A fails.

In other words, too-small return buffers is a failure mode of calls that are completely ignored by the summaries: from the perspective of the caller, the call external is summarized in such a way that the call "appears" to have exactly returned the amount of data required.

The optimistic return size is a flag to avoid insufficient returndata reverts (presumably when running with `@withrevert`). Like the summaries above, with the optimisticReturnsize flag, any unresolved (havocing) function will still be assumed to return the necessary number of bytes (even if those bytes are completely arbitrary).

File: docs/ref-manual/cvl/methods.md (hunk 7 above)
Comment (jtoman) [reply docs/ref-manual/cvl/methods.md:290]: external library calls are delegate calls, and by the yellow paper they cannot (and should not) transfer balances. Internal library calls are not message calls, and so do not directly change balances (unless the body of these functions have external, balance transferring calls)

File: docs/ref-manual/cvl/methods.md [hunk 9]
```diff
@@ -46,44 +55,280 @@ method_summary   ::= "ALWAYS" "(" value ")"
...
+   unchanged.
+
+```{todo}
+The effect of library calls on the current contract's ETH balance is unclear.
+
+The following comment from the old docs should be clarified:
+
+If the contract you are verifying relies heavily on modification of ETH
+balances, it's recommended to identify the balance-modifying functions and mark
+them `HAVOC_ALL` if necessary.
```
Comment (jtoman) [inline docs/ref-manual/cvl/methods.md:296]: This is pretty weird advice, I have no idea

**Review by jtoman — COMMENTED** [review]
sorry for the late turn around

File: docs/ref-manual/cvl/methods.md (hunk 1 above)
Comment (mdgeorge4153) [reply docs/ref-manual/cvl/methods.md:65]: ok, I removed expression and blocks and with env, leaving only the function call syntax.  The section describing function summaries only lists CVL functions and ghosts as possiblities.

File: docs/ref-manual/cvl/methods.md (hunk 3 above)
Comment (mdgeorge4153) [reply docs/ref-manual/cvl/methods.md:152]: fixed.

File: docs/ref-manual/cvl/methods.md (hunk 4 above)
Comment (mdgeorge4153) [reply docs/ref-manual/cvl/methods.md:156]: fixed.

File: docs/ref-manual/cvl/methods.md (hunk 5 above)
Comment (mdgeorge4153) [reply docs/ref-manual/cvl/methods.md:193]: I haven't run any tests, but I think this behavior is what I would expect so I will document it so and consider deviations from it to be bugs.

File: docs/ref-manual/cvl/methods.md (hunk 7 above)
Comment (mdgeorge4153) [reply docs/ref-manual/cvl/methods.md:290]: I replaced the third bullet by "Calls to library methods and `delegatecall`s are assumed to change the caller's storage in an arbitrary way, but are assumed to leave ETH balances and the storage of other contracts unchanged."  Is this accurate now?

File: docs/ref-manual/cvl/methods.md (hunk 8 above)
Comment (mdgeorge4153) [reply docs/ref-manual/cvl/methods.md:330]: removed

File: docs/ref-manual/cvl/methods.md (hunk 2 above)
Comment (mdgeorge4153) [reply docs/ref-manual/cvl/methods.md]: I understand now - payable functions depend on the msg.value which is part of the environment.  Updated.

File: docs/ref-manual/cvl/methods.md (hunk 6 above)
Comment (mdgeorge4153) [reply docs/ref-manual/cvl/methods.md]: I think I have accurately described the situation.

File: docs/ref-manual/cvl/methods.md (hunk 7 above)
Comment (mdgeorge4153) [reply docs/ref-manual/cvl/methods.md]: I think I've accurately updated the documentation

File: docs/confluence/advanced/methods.md [hunk 10]
```diff
@@ -1,11 +1,228 @@
...
+
+Internal Function Summaries
+===========================
+
+Summaries Always Inserted
+-------------------------
+
+Summaries for **external** functions are only inserted when an implementation
+of that function cannot be found, and so we default to some summary that we,
+the verifier, sees fit. _However_, internal functions can always be found and
```
Comment (CowardlyBattleCat) [inline docs/confluence/advanced/methods.md:24]: For clarity and grammar, I would change the beginning of line 24 to "...as the verifier, see fit."

File: docs/confluence/advanced/methods.md [hunk 11]
```diff
@@ -1,11 +1,228 @@
...
+```cvl
+methods {
+  mul(uint256 x, uint256 y) => easier_multiplication(x, y);
+}
+```
+
+Note this specific summarization is very dangerous and may cause vacuity bugs.
+
+In simple cases, these summaries may be used to replace harnesses, though the
+fact that they cannot call contract functions limits the types of harnesses
```
Comment (jtoman) [inline docs/confluence/advanced/methods.md:227]: ?? summaries?

File: docs/ref-manual/cvl/methods.md [hunk 12]
```diff
@@ -46,44 +55,251 @@ method_summary   ::= "ALWAYS" "(" value ")"
...
+
+The application policy determines which function calls are replaced by
+approximations.  See {ref}`summaries` for details.
+
+(summaries)=
+Which function calls are summarized
+-----------------------------------
+
+Method summaries apply to all calls with the matching ABI signature, including
+internal methods and external methods on all contracts.  There is currently no
```
Comment (jtoman) [inline docs/ref-manual/cvl/methods.md:125]: nit: internal methods don't have an ABI signature, and in point of fact, the signatures that we write in the methods block aren't (necessarily) ABI signatures.

File: docs/ref-manual/cvl/methods.md [hunk 13]
```diff
@@ -46,44 +55,251 @@ method_summary   ::= "ALWAYS" "(" value ")"
...
+way to apply different summaries to different contracts or to summarize some
+calls and not others to methods with the same ABI signature.  For this reason,
```
Comment (jtoman) [inline docs/ref-manual/cvl/methods.md:127]: this is not, strictly speaking, true. It is possible that for some function, e.g., `owner()`, the callee for some call-sites is known and for others it is not. If you don't specify an forced summary, then the summary behavior will only apply for those callsites where the callee is unresolved.

File: docs/ref-manual/cvl/methods.md [hunk 14]
```diff
@@ -46,44 +55,251 @@ method_summary   ::= "ALWAYS" "(" value ")"
...
+
+The Solidity compiler allows internal functions to have richer type
+specifications than just ABI specifications, but the prover only considers the
+ABI encoding to determine whether a method call should be summarized.  For
+example, internal functions can accept struct arguments, but the structs are
+replaced by tuples in the ABI signature.  If the generated ABI signature of an
+internal method call matches the ABI signature specified in the methods block,
+it will be a candidate for summarization even if the declared argument types in
+the Solidity function don't exactly match the declared argument types in the
+`methods` block.
```
Comment (jtoman) [inline docs/ref-manual/cvl/methods.md:139]: @thomasbernardi I thought this was no longer the case

File: docs/ref-manual/cvl/methods.md [hunk 15]
```diff
@@ -46,44 +55,251 @@ method_summary   ::= "ALWAYS" "(" value ")"
...
+ * With `DISPATCHER(false)` or just `DISPATCHER`, the unknown contract is
+   considered as a possiblity
+
+ * With `DISPATCHER(true)`, only the known contract instances are considered
+
+The most commonly used option is `DISPATCHER(true)`, because in most cases the
+behavior of `DISPATCHER(false)` is equivalent to that of `AUTO`.
+`DISPATCHER(false)` only behaves differently from `AUTO` when one of the known
+contracts exhibits behavior that `AUTO` assumes is impossible (such as
+reentrancy).
```
Comment (jtoman) [inline docs/ref-manual/cvl/methods.md:262]: I don't think this is true? If the callee address is *always* one of the known contracts (as proven by z3) then the two dispatcher flavors are equivalent? Maybe you're making a point about possible static behaviors...

File: docs/confluence/advanced/methods.md (hunk 11 above)
Comment (mdgeorge4153) [reply docs/confluence/advanced/methods.md:227]: same here

File: docs/ref-manual/cvl/methods.md (hunk 12 above)
Comment (mdgeorge4153) [reply docs/ref-manual/cvl/methods.md:125]: my understanding was that we take the signature of the internal method or the methods block, erase all of the information that goes beyond what would be encoded in an ABI signature, and then use that for dispatching.  Is that incorrect?  It sounds like either way this needs clarification

File: docs/ref-manual/cvl/methods.md (hunk 13 above)
Comment (mdgeorge4153) [reply docs/ref-manual/cvl/methods.md:127]: I think this behavior is described below, but I should be more careful here to distinguish between which methods match and which methods are approximated.

File: docs/ref-manual/cvl/methods.md (hunk 14 above)
Comment (mdgeorge4153) [reply docs/ref-manual/cvl/methods.md:139]: We should discuss and maybe look at tests, because this was my understanding of what Thomas explained to me

File: docs/ref-manual/cvl/methods.md (hunk 13 above)
Comment (thomasbernardi) [reply docs/ref-manual/cvl/methods.md:127]: > this is not, strictly speaking, true. It is possible that for some function, e.g., owner(), the callee for some call-sites is known and for others it is not. If you don't specify an forced summary, then the summary behavior will only apply for those callsites where the callee is unresolved.

I think this is not quite what we're talking about here. Perhaps there should be a more clear distinction in language between how summaries are "specified" and how the tool "applies" them.

File: docs/ref-manual/cvl/methods.md (hunk 12 above)
Comment (thomasbernardi) [reply docs/ref-manual/cvl/methods.md:125]: Internal method's _shouldn't_ have an ABI signature and as far as `solc` is concerned they do not. However, as far as our tool is concerned, all methods have an ABI signature. Maybe we can use some language like "the equivalent ABI signature" or isomorphic or something.

Comment (mdgeorge4153) [reply docs/ref-manual/cvl/methods.md:139]: I think the latest draft clarifies this

File: docs/ref-manual/cvl/methods.md (hunk 13 above)
Comment (mdgeorge4153) [reply docs/ref-manual/cvl/methods.md:141]: I think the latest draft clarifies this

File: docs/ref-manual/cvl/methods.md (hunk 14 above)
Comment (mdgeorge4153) [reply docs/ref-manual/cvl/methods.md]: resolved in latest draft

File: docs/ref-manual/cvl/methods.md (hunk 15 above)
Comment (mdgeorge4153) [reply docs/ref-manual/cvl/methods.md]: I weakened the claim in the current draft to something that I now think is true.

**Review by nd-certora — APPROVED** [review]

File: docs/ref-manual/cvl/methods.md [hunk 16]
```diff
@@ -46,44 +55,251 @@ method_summary   ::= "ALWAYS" "(" value ")"
                    | "HAVOC_ALL"
                    | "DISPATCHER" [ "(" ( "true" | "false" ) ")" ]
                    | "AUTO"
-                   | [ "with" "(" cvl_params ")" ] block
-                   | [ "with" "(" cvl_params ")" ] expression
-
-cvl_param ::= cvl_type [ id ]
+                   | id "(" [ id { "," id } ] ")"
 
 ```
```
[the remark this answers is not in the document]
Comment (mdgeorge4153) [reply docs/ref-manual/cvl/methods.md:61]: I think this is a good idea except that it will lead to a lot of duplication.  This fact is included in the description below though.

File: docs/ref-manual/cvl/methods.md [hunk 17]
```diff
@@ -1,21 +1,30 @@
...
+* **Summary declarations** are used to replace calls to methods having the
   given signature with something that is simpler for the Prover to reason about.
   Summaries allow the Prover to reason about external contracts whose code is
   unavailable.  They can also be useful to simplify the code being verified to
   circumvent timeouts.
 
+```{caution}
+Summary declarations change the way that some function calls are interpreted,
+and are therefore {term}`unsound` (with the exception of `HAVOC_ALL` summaries,
+which are always sound).
```
[the remark this answers is not in the document]
Comment (mdgeorge4153) [reply docs/ref-manual/cvl/methods.md:25]: fixed

File: docs/ref-manual/cvl/methods.md [hunk 18]
```diff
@@ -46,44 +55,251 @@ method_summary   ::= "ALWAYS" "(" value ")"
...
+
+ * {ref}`const-summary`
+ * {ref}`havoc-summary`
+ * {ref}`dispatcher`
+ * {ref}`auto-summary`
+ * {ref}`function-summary`
+
+The application policy determines which function calls are replaced by
+approximations.  See {ref}`summaries` for details.
+
```
[the remark this answers is not in the document]
Comment (mdgeorge4153) [reply docs/ref-manual/cvl/methods.md:119]: @thomasbernardi ?

File: docs/confluence/advanced/methods.md [hunk 19]
```diff
@@ -1,11 +1,228 @@
...
+following summaries are allowed:
+
+*   `ALWAYS(X)` the summary always returns `X` and has no side-effects
+    
+*   `CONSTANT` the summary always returns the same constant and has no side
+    effects
+    
+*   `NONDET` the summary returns a havoc'd value
+    
+*   `Ghost` the summary returns the value return by the given ghost function
```
Comment (thomasbernardi) [inline docs/confluence/advanced/methods.md:50]: CVL Function should be allowed too right?

**Comment by mdgeorge4153** [comment]
We should update the pointer in the “Prover approximations” section to refer here.  We should also remove more of the old docs.

File: docs/confluence/advanced/methods.md (hunk 19 above)
Comment (mdgeorge4153) [reply docs/confluence/advanced/methods.md:50]: confluence docs are deprecated.

File: docs/ref-manual/cvl/methods.md (hunk 18 above)
[the remark this answers is not in the document]
Comment (mdgeorge4153) [reply docs/ref-manual/cvl/methods.md:120]: Thomas is designing a test suite to figure out exactly what the limitations are; in the mean time I have added the limitations  I know about and left a TODO that there are other unsupported combinations.  Once we have more accurate information I'll submit another PR.
